### PR TITLE
[SPARK-38415][SQL] Update the histogram_numeric (x, y) result type to make x == the input type

### PR DESCRIPTION
**Pull request by dtenedor**
### What changes were proposed in this pull request?

This pull request updates the histogram_numeric SQL function to support more numeric input types, returning the results an an array of structs of two fields each. The first field has the same type as the first argument to the histogram_numeric aggregate function (rather than always having double type before this change). This removes the need for the user to apply a cast function to the result in order to use it.

Example behavior after this change becomes effective:

SELECT histogram_numeric(col, 3) FROM VALUES (TIMESTAMP '2017-03-01 00:00:00'),
(TIMESTAMP '2017-04-01 00:00:00'), (TIMESTAMP '2017-05-01 00:00:00') AS tab(col);

Returns type: struct<histogram_numeric(col, 3):array<struct<x:timestamp,y:double>>>.
Query output: [{"x":2017-03-01 00:00:00,"y":1.0},{"x":2017-04-01 00:00:00,"y":1.0},{"x":2017-05-01 00:00:00,"y":1.0}].

### Why are the changes needed?

This removes the need for users to explicitly cast the function result type in many cases.

### Does this PR introduce _any_ user-facing change?

Yes, it changes the `histogram_numeric` function result type.

### How was this patch tested?

Unit tests, file-based query tests.

### CODE REVIEW

File: sql/core/src/test/resources/sql-tests/results/group-by.sql.out [hunk 1]
```diff
@@ -708,9 +708,112 @@ FROM VALUES
  (31), (32), (33), (34), (35), (3), (37), (38), (39), (40),
  (41), (42), (43), (44), (45), (46), (47), (48), (49), (50) AS tab(col)
 -- !query schema
-struct<histogram_2:array<struct<x:double,y:double>>,histogram_3:array<struct<x:double,y:double>>,histogram_5:array<struct<x:double,y:double>>,histogram_10:array<struct<x:double,y:double>>>
+struct<histogram_2:array<struct<x:int,y:double>>,histogram_3:array<struct<x:int,y:double>>,histogram_5:array<struct<x:int,y:double>>,histogram_10:array<struct<x:int,y:double>>>
```
Comment (HyukjinKwon) [inline sql/core/src/test/resources/sql-tests/results/group-by.sql.out:711]: We should at least update the migration guide (https://github.com/apache/spark/blob/master/docs/sql-migration-guide.md) or even have a flag. cc @sigmod FYI

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 2]
```diff
@@ -4294,6 +4294,37 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
         Row(3, 2, 6) :: Nil)
     }
   }
+
+  test("SC-93338: Cover the histogram_numeric aggregate function for many input types") {
```
Comment (HyukjinKwon) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4298]:
```suggestion
  test("SPARK-38415: Cover the histogram_numeric aggregate function for many input types") {
```

File: sql/core/src/test/resources/sql-tests/results/group-by.sql.out (hunk 1 above)
Comment (HyukjinKwon) [reply sql/core/src/test/resources/sql-tests/results/group-by.sql.out:711]: I guess the double value behaivour is from [Hive's](https://cwiki.apache.org/confluence/display/hive/languagemanual+udf) while I like it to match the input type. I think we should have a switch to turn on and off this behaviour.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumeric.scala [hunk 3]
```diff
@@ -125,7 +125,21 @@ case class HistogramNumeric(
...
         val coord = buffer.getBin(index)
-        InternalRow.apply(coord.x, coord.y)
+        // The 'coord.x' and 'coord.y' have double type here. If the expression is expected to
+        // return a result of double type, this is sufficient. Otherwise, we need to internally
+        // convert these values to the expected result type, for cases like timestamps and intervals
+        // which are valid inputs to the numeric histogram aggregate function.
+        val result: Any = left.dataType match {
+          case ByteType => coord.x.toByte
+          case IntegerType | DateType | _: YearMonthIntervalType =>
+            coord.x.toInt
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumeric.scala:135]: I think we should explain this behaviour in the documentation above L54 with an example.

File: sql/core/src/test/resources/sql-tests/results/group-by.sql.out (hunk 1 above)
Comment (cloud-fan) [reply sql/core/src/test/resources/sql-tests/results/group-by.sql.out:711]: Makes sense. I think we can follow the `Size` expression, which has a legacy config to restore to the old buggy behavior.

Comment (dtenedor) [reply sql/core/src/test/resources/sql-tests/results/group-by.sql.out:711]: Thanks! I've added the suggested flag to toggle this behavior. This seems like a reasonable config to have.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 4]
```diff
@@ -3562,6 +3562,18 @@ object SQLConf {
...
 
+  val HISTOGRAM_NUMERIC_PROPAGATE_INPUT_TYPE =
+    buildConf("spark.sql.legacy.histogramNumericPropagateInputType")
+      .internal()
+      .doc("The histogram_numeric function computes a histogram on numeric 'expr' using nb bins. " +
+        "The return value is an array of (x,y) pairs representing the centers of the histogram's " +
+        "bins. If this config is set to true, the output type of the 'x' field in the return " +
+        "value is propagated from the input value consumed in the aggregate function. Otherwise, " +
+        "'x' always has double type.")
+      .version("3.4.0")
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3573]:
```suggestion
      .version("3.3.0")
```

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 2 above)
Comment (HyukjinKwon) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4332]: @d3borah seems like this was missed :-).

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala [hunk 5]
```diff
@@ -3562,6 +3562,18 @@ object SQLConf {
...
+      .booleanConf
+      .createWithDefault(true)
```
Comment (HyukjinKwon) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3575]: If this gets enabled by default, let's note this in the migration guide.

File: sql/core/src/test/resources/sql-tests/inputs/group-by.sql [hunk 6]
```diff
@@ -211,6 +211,27 @@ FROM VALUES
...
+SELECT histogram_numeric(col, 3) FROM VALUES
+  (CAST(1 AS SMALLINT)), (CAST(2 AS SMALLINT)), (CAST(3 AS SMALLINT)) AS tab(col);
+SELECT histogram_numeric(col, 3) FROM VALUES
+  (CAST(1 AS BIGINT)), (CAST(2 AS BIGINT)), (CAST(3 AS BIGINT)) AS tab(col);
+SELECT histogram_numeric(col, 3) FROM VALUES (TIMESTAMP '2017-03-01 00:00:00'),
+  (TIMESTAMP '2017-04-01 00:00:00'), (TIMESTAMP '2017-05-01 00:00:00') AS tab(col);
+SELECT histogram_numeric(col, 3) FROM VALUES (INTERVAL '100-00' YEAR TO MONTH),
+  (INTERVAL '110-00' YEAR TO MONTH), (INTERVAL '120-00' YEAR TO MONTH) AS tab(col);
+SELECT histogram_numeric(col, 3) FROM VALUES (INTERVAL '12 20:4:0' DAY TO SECOND),
+  (INTERVAL '12 21:4:0' DAY TO SECOND), (INTERVAL '12 22:4:0' DAY TO SECOND) AS tab(col);
```
Comment (sigmod) [inline sql/core/src/test/resources/sql-tests/inputs/group-by.sql:233]: Does it make sense to add the following test query if it doesn't exist elsewhere?

SELECT histogram_numeric(col, 3) FROM VALUES (null), (null), (null) AS tab(col);  -- x type is `void`?

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 5 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3575]: Done.

File: sql/core/src/test/resources/sql-tests/inputs/group-by.sql (hunk 6 above)
Comment (dtenedor) [reply sql/core/src/test/resources/sql-tests/inputs/group-by.sql:233]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumeric.scala [hunk 7]
```diff
@@ -46,7 +47,10 @@ import org.apache.spark.sql.util.NumericHistogram
       smaller datasets. Note that this function creates a histogram with non-uniform
       bin widths. It offers no guarantees in terms of the mean-squared-error of the
       histogram, but in practice is comparable to the histograms produced by the R/S-Plus
-      statistical computing packages.
+      statistical computing packages. Note: if
+      SQLConf.spark.sql.legacy.histogramNumericPropagateInputType is set to true, the output type
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumeric.scala:51]: Let's not mention the legacy config as it's internal and only for users who have problems migrating to new Spark versions. Let's just document the new behavior.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 4 above)
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3573]: hmm, do we want to merge this PR after 2 weeks to not include it in Spark 3.3.0? Otherwise I think the version should be `3.3.0`

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumericSuite.scala [hunk 8]
```diff
@@ -52,7 +57,8 @@ class HistogramNumericSuite extends SparkFunSuite {
     assert(compareEquals(buffer,
       NumericHistogramSerializer.deserialize(NumericHistogramSerializer.serialize(buffer))))
 
-    val agg = new HistogramNumeric(BoundReference(0, DoubleType, true), Literal(5))
+    val agg = new HistogramNumeric(
+      BoundReference(0, DoubleType, true), Literal(5))
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumericSuite.scala:61]: nit: unnecessary change.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumericSuite.scala [hunk 9]
```diff
@@ -87,7 +92,8 @@ class HistogramNumericSuite extends SparkFunSuite {
 
   test("class HistogramNumeric, fails analysis if nBins is not a constant") {
     val attribute = AttributeReference("a", IntegerType)()
-    val wrongNB = new HistogramNumeric(attribute, nBins = AttributeReference("b", IntegerType)())
+    val wrongNB = new HistogramNumeric(
+      attribute, nBins = AttributeReference("b", IntegerType)())
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumericSuite.scala:96]: ditto

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala [hunk 10]
```diff
@@ -4328,6 +4328,37 @@ class SQLQuerySuite extends QueryTest with SharedSparkSession with AdaptiveSpark
...
+        (INTERVAL '110-00' YEAR TO MONTH), (INTERVAL '120-00' YEAR TO MONTH) AS tab(col)""",
+      """SELECT histogram_numeric(col, 3) FROM VALUES (INTERVAL '12 20:4:0' DAY TO SECOND),
+        (INTERVAL '12 21:4:0' DAY TO SECOND), (INTERVAL '12 22:4:0' DAY TO SECOND) AS tab(col)"""
+    )) {
+      val df = sql(query)
+      val result = df.collect.head
+      // As a basic sanity check, ensure that the result is not NULL. This makes sure that we run
+      // all stages of the aggregate function and evaluate the result to some concrete value each
+      // time.
+      assert(!result.isNullAt(0))
```
Comment (cloud-fan) [inline sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4359]: It seems to me that end-to-end tests are fully covered by `group-by.sql` which checks the result not just nullability, and we can remove this test case.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumeric.scala (hunk 7 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumeric.scala:51]: Done.

File: sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala (hunk 4 above)
Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/internal/SQLConf.scala:3573]: Sorry, this should be 3.3.0. Fixed.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumericSuite.scala (hunk 8 above)
Comment (dtenedor) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumericSuite.scala:61]: Done.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumericSuite.scala (hunk 9 above)
Comment (dtenedor) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumericSuite.scala:96]: Done.

File: sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala (hunk 10 above)
Comment (dtenedor) [reply sql/core/src/test/scala/org/apache/spark/sql/SQLQuerySuite.scala:4359]: Good point, done.

File: sql/core/src/test/resources/sql-tests/results/group-by.sql.out [hunk 11]
```diff
@@ -708,9 +708,121 @@ FROM VALUES
...
+struct<histogram_numeric(col, 3):array<struct<x:interval day to second,y:double>>>
+-- !query output
+[{"x":12 20:04:00.000000000,"y":1.0},{"x":12 21:04:00.000000000,"y":1.0},{"x":12 22:04:00.000000000,"y":1.0}]
+
+
+-- !query
+SELECT histogram_numeric(col, 3)
+FROM VALUES (null), (null), (null) AS tab(col)
+-- !query schema
+struct<histogram_numeric(col, 3):array<struct<x:double,y:double>>>
```
Comment (sigmod) [inline sql/core/src/test/resources/sql-tests/results/group-by.sql.out:823]: Ah, `x` it's still `double`.  Is it intended?

Comment (dtenedor) [reply sql/core/src/test/resources/sql-tests/results/group-by.sql.out:823]: Yeah, looks like the `null` is interpreted as type `double`. I added a couple more test cases with explicit `cast(null as <type>)` to double-check.

Comment (sigmod) [reply sql/core/src/test/resources/sql-tests/results/group-by.sql.out:823]: Ok, this is what I get:

SELECT typeof(col) FROM VALUES (null), (null) AS tab(col)
-
void
void

I was hoping x becomes a `void` with this PR :-)
But I'm fine with `double` if you think that's an intended behavior of `histogram_numeric`.

**Review by sigmod — APPROVED** [review]

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumeric.scala [hunk 12]
```diff
@@ -124,8 +128,33 @@ case class HistogramNumeric(
...
+            case ByteType => coord.x.toByte
+            case IntegerType | DateType | _: YearMonthIntervalType =>
+              coord.x.toInt
+            case FloatType => coord.x.toFloat
+            case ShortType => coord.x.toShort
+            case _: DayTimeIntervalType | LongType | TimestampType | TimestampNTZType =>
+              coord.x.toLong
+            case _ => coord.x
+          }
+          InternalRow.apply(result, coord.y)
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumeric.scala:150]: After some more thoughts, maybe we can simply reuse `Cast` here
```
InternalRow.apply(Cast(Literal(coord.x, DoubleType), left.dataType).eval(), coord.y)
```
Can we give it a try? It can simplify the code.

**Review by cloud-fan — APPROVED** [review]

**Review by dtenedor — COMMENTED** [review]
Thanks all again for your reviews!

**Comment by dtenedor** [comment]
(Looks like I accidentally re-requested review by mistake, but this seems done/ready to merge if you guys agree.)

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumeric.scala [hunk 13]
```diff
@@ -72,6 +74,8 @@ case class HistogramNumeric(
     case n: Int => n
   }
 
+  private def propagateInputType: Boolean = SQLConf.get.histogramNumericPropagateInputType
```
Comment (cloud-fan) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumeric.scala:77]: This is another tricky thing: the semantic/behavior of an expression should be static, not dynamic that can be changed in the middle at runtime. Think about a user who creates a view with this function, then he/she read the view back, the behavior should be the same with when the view was created, no matter what the config is currently.

That's why we have `case class Size(child: Expression, legacySizeOfNull: Boolean)`: we put the flag in constructor, so that the semantic/behavior is fixed once the expression is created.

Comment (dtenedor) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/expressions/aggregate/HistogramNumeric.scala]: Thanks, that's a good point. Updated to make this static.

File: docs/sql-migration-guide.md [hunk 14]
```diff
@@ -24,6 +24,8 @@ license: |
 
 ## Upgrading from Spark SQL 3.2 to 3.3
 
+  - Since Spark 3.3, the `histogram_numeric` function in Spark SQL returns an output type of an array of structs (x, y), where the type of the 'x' field in the return value is propagated from the input values consumed in the aggregate function. In previous versions, 'x' always had double type.
```
Comment (HyukjinKwon) [inline docs/sql-migration-guide.md:27]: Can we mention which configuration should be enabled to restore the previous behaviour?

**Review by HyukjinKwon — APPROVED** [review]
LGTM2 otherwise

File: docs/sql-migration-guide.md (hunk 14 above)
Comment (dtenedor) [reply docs/sql-migration-guide.md]: Sure, this is done.

File: docs/sql-migration-guide.md [hunk 15]
```diff
@@ -24,6 +24,8 @@ license: |
 
 ## Upgrading from Spark SQL 3.2 to 3.3
 
+  - Since Spark 3.3, the `histogram_numeric` function in Spark SQL returns an output type of an array of structs (x, y), where the type of the 'x' field in the return value is propagated from the input values consumed in the aggregate function. In Spark 3.2 or earlier, 'x' always had double type. Optionally, use the configuration SQLConf.spark.sql.legacy.histogramNumericPropagateInputType since Spark 3.3 to revert back to the previous behavior.
```
Comment (HyukjinKwon) [inline docs/sql-migration-guide.md:27]:
```suggestion
  - Since Spark 3.3, the `histogram_numeric` function in Spark SQL returns an output type of an array of structs (x, y), where the type of the 'x' field in the return value is propagated from the input values consumed in the aggregate function. In Spark 3.2 or earlier, 'x' always had double type. Optionally, use the configuration `spark.sql.legacy.histogramNumericPropagateInputType` since Spark 3.3 to revert back to the previous behavior. 
```

**Comment by cloud-fan** [comment]
thanks, merging to master!
